[WPE] WPE Platform: do not release buffers when rendered

carlosgcampos opened the pull request:
#### 01716ada465537d722308dd31ec31171ffeffe86
<pre>
[WPE] WPE Platform: do not release buffers when rendered
<a href="https://bugs.webkit.org/show_bug.cgi?id=270693">https://bugs.webkit.org/show_bug.cgi?id=270693</a>

Reviewed by Alejandro G. Castro.

The buffers can be rendered again, so we need to keep them locked until
a new frame is ready to be rendered. This patch adds
WPEView::buffer-released in addition to the existing
WPEView::buffer-rendered to notify that the buffer is no longer used and
can be destroyed or reused. This patch also changes the WPEBuffer API to
receive a WPEView in constructor instead of WPEDisplay, since buffers
are actually attached to a specific WPEView, because we need to easily
get the WPEView from the wayland buffer release callback.

* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::AcceleratedBackingStoreDMABuf):
(WebKit::AcceleratedBackingStoreDMABuf::updateSurfaceID):
(WebKit::AcceleratedBackingStoreDMABuf::didCreateBuffer):
(WebKit::AcceleratedBackingStoreDMABuf::didCreateBufferSHM):
(WebKit::AcceleratedBackingStoreDMABuf::bufferRendered):
(WebKit::AcceleratedBackingStoreDMABuf::bufferReleased):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/WPEPlatform/wpe/WPEBuffer.cpp:
(wpeBufferGetProperty):
(wpe_buffer_class_init):
(wpe_buffer_get_view):
(wpe_buffer_get_display): Deleted.
* Source/WebKit/WPEPlatform/wpe/WPEBuffer.h:
* Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.cpp:
(wpeBufferDMABufDispose):
(wpeBufferDMABufImportToEGLImage):
(wpeBufferDMABufTryEnsureGBMDevice):
(wpe_buffer_dma_buf_new):
* Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.h:
* Source/WebKit/WPEPlatform/wpe/WPEBufferSHM.cpp:
(wpe_buffer_shm_new):
* Source/WebKit/WPEPlatform/wpe/WPEBufferSHM.h:
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_class_init):
(wpe_view_buffer_released):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/drm/WPEViewDRM.cpp:
(drmBufferCreateDMABuf):
(drmBufferCreate):
(wpeViewDRMRequestUpdate):
(wpeViewDRMRenderBuffer):
(wpeViewDRMDidPageFlip):
* Source/WebKit/WPEPlatform/wpe/headless/WPEViewHeadless.cpp:
(wpeViewHeadlessRenderBuffer):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(createWaylandBufferFromDMABuf):
(createWaylandBufferSHM):
(createWaylandBuffer):
(createWaylandBufferFromEGLImage): Deleted.

Canonical link: <a href="https://commits.webkit.org/275903@main">https://commits.webkit.org/275903@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36ffe27e4ab96acf35126b9cf1319daa43f34cc3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43010 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22036 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45412 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45634 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39135 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25774 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19459 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35567 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43583 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19082 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37036 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16562 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16673 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1066 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39204 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38425 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47166 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17857 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14713 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42342 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19462 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40995 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19641 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5867 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19092 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alexgcastro reviewed (approved):
Just validated the change and it works as expected, thanks!

carlosgcampos committed:
[WPE] WPE Platform: do not release buffers when rendered
https://bugs.webkit.org/show_bug.cgi?id=270693

Reviewed by Alejandro G. Castro.

The buffers can be rendered again, so we need to keep them locked until
a new frame is ready to be rendered. This patch adds
WPEView::buffer-released in addition to the existing
WPEView::buffer-rendered to notify that the buffer is no longer used and
can be destroyed or reused. This patch also changes the WPEBuffer API to
receive a WPEView in constructor instead of WPEDisplay, since buffers
are actually attached to a specific WPEView, because we need to easily
get the WPEView from the wayland buffer release callback.

* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::AcceleratedBackingStoreDMABuf):
(WebKit::AcceleratedBackingStoreDMABuf::updateSurfaceID):
(WebKit::AcceleratedBackingStoreDMABuf::didCreateBuffer):
(WebKit::AcceleratedBackingStoreDMABuf::didCreateBufferSHM):
(WebKit::AcceleratedBackingStoreDMABuf::bufferRendered):
(WebKit::AcceleratedBackingStoreDMABuf::bufferReleased):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/WPEPlatform/wpe/WPEBuffer.cpp:
(wpeBufferGetProperty):
(wpe_buffer_class_init):
(wpe_buffer_get_view):
(wpe_buffer_get_display): Deleted.
* Source/WebKit/WPEPlatform/wpe/WPEBuffer.h:
* Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.cpp:
(wpeBufferDMABufDispose):
(wpeBufferDMABufImportToEGLImage):
(wpeBufferDMABufTryEnsureGBMDevice):
(wpe_buffer_dma_buf_new):
* Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.h:
* Source/WebKit/WPEPlatform/wpe/WPEBufferSHM.cpp:
(wpe_buffer_shm_new):
* Source/WebKit/WPEPlatform/wpe/WPEBufferSHM.h:
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_class_init):
(wpe_view_buffer_released):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/drm/WPEViewDRM.cpp:
(drmBufferCreateDMABuf):
(drmBufferCreate):
(wpeViewDRMRequestUpdate):
(wpeViewDRMRenderBuffer):
(wpeViewDRMDidPageFlip):
* Source/WebKit/WPEPlatform/wpe/headless/WPEViewHeadless.cpp:
(wpeViewHeadlessRenderBuffer):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(createWaylandBufferFromDMABuf):
(createWaylandBufferSHM):
(createWaylandBuffer):
(createWaylandBufferFromEGLImage): Deleted.

Canonical link: https://commits.webkit.org/275903@main
